shared/utility/math/filter/ParticleFilter.hpp

```diff
@@ -38,242 +74,393 @@ namespace utility::math::filter {
-            // Expectation Maximisation. International Conference on Signal Processing and Communication
-            // Systems, 2017.
-
-            ParticleList new_particles =
-                ParticleList::NullaryExpr(Model::size, n_particles, [&]() { return norm(rng); });
-
-            const StateMat sqrt_covariance = covariance.cwiseSqrt();
+        /**
+         * @brief Constructs the particle filter using default values. Mean is set to 0 and covariance is set to
+         * \f$0.1 * I_{n}\f$, where n is the size of the model. The residual method is used for resampling particles and
+         * the residual will use the systematic method for resampling the residual particles.
+         */
+        ParticleFilter() {
+            set_state(StateVec::Zero(), StateMat::Identity() * 0.1);
+            resample_method = ResampleMethod::RESIDUAL;
+            residual_method = ResampleMethod::SYSTEMATIC;
+        }
 
-            for (int i = 0; i < n_particles; ++i) {
-                new_particles.col(i) = sqrt_covariance * new_particles.col(i) + mean;
+        /**
+         * @brief Constructs the particle filter with the provided values. The resampling method with default to the
+         * residual method with the systematic method for resampling the residual particles. It is an error to use the
+         * residual method for resampling residuals and an exception will be thrown.
+         *
+         * @param mean The mean of the multivariate normal distribution. Used for sampling the initial particles in the
+         * filter.
+         * @param covariance The covariance of the multivariate normal distribution. Used for sampling the initial
+         * particles in the filter.
+         * @param resample_method_ The method to use when resampling particles during the time update step.
+         * @param residual_method_ When using ResampleMethod::RESIDUAL for the resample method, this method will be used
+         * when the residual method needs to resample residual particles.
+         */
+        ParticleFilter(const StateVec& mean,
+                       const StateMat& covariance,
+                       const ResampleMethod& resample_method_ = ResampleMethod::RESIDUAL,
+                       const ResampleMethod& residual_method_ = ResampleMethod::SYSTEMATIC) {
+            set_state(mean, covariance);
+            resample_method = residual_method_;
+
+            // The residual resampling method requires a secondary method to resample the residual particles. This
+            // secondary method cannot be the residual method.
+            if (residual_method == ResampleMethod::RESIDUAL) {
+                throw std::runtime_error(
+                    "Invalid setting for residual method. Can't use the residual resampling method to sample "
+                    "residuals.");
             }
-
-            return new_particles;
+            residual_method = resample_method_;
```

Comment:
I think you need to not store a in b and b in a...

Reply:
Lol @Bidski

Reply:
Lol @TrentHouliston you missed it too